File: Graph/NumberOfIslands.java
package Graph;

import java.util.LinkedList;
import java.util.Queue;

public class NumberOfIslands {
    public boolean isValid(int x, int y, char[][] grid) {
        if (x >= 0 && y >= 0 && x < grid.length && y < grid[0].length && grid[x][y] == '1')
            return true;
        return false;
    }

    public int numIslands(char[][] grid) {
        // Array to store the four directions: up, down, right, left
        int[][] directions = new int[][] { { -1, 0 }, { 1, 0 }, { 0, 1 }, { 0, -1 } };

        int islands = 0; // Counter for the number of islands
        int n = grid.length; // Number of rows in the grid
        int m = grid[0].length; // Number of columns in the grid

        // Nested loops to iterate through each cell in the grid
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                // If the cell contains land (represented by '1'), perform a breadth-first
                // search
                if (grid[i][j] == '1') {
                    bfs(i, j, grid, directions); // Call the breadth-first search function
                    islands++; // Increment the island counter
                }
            }
        }

        return islands; // Return the total number of islands
    }

    public void bfs(int i, int j, char[][] grid, int[][] directions) {
        // Initialize a queue to store the indices of the cells to be visited
        Queue<int[]> q = new LinkedList<>();
        // Add the initial cell indices to the queue
        q.offer(new int[] { i, j });
        // Process cells in the queue until it's empty
        while (!q.isEmpty()) {
            // Retrieve and remove the cell indices from the front of the queue
            int[] temp = q.poll();
            // Explore all neighboring cells based on the given directions
            for (int[] dir : directions) {
                int x = temp[0] + dir[0]; // Calculate the x-coordinate of the neighboring cell
                int y = temp[1] + dir[1]; // Calculate the y-coordinate of the neighboring cell
                // Check if the neighboring cell is within the grid and is valid
                if (isValid(x, y, grid)) {
                    // Add the neighboring cell to the queue for further exploration
                    q.offer(new int[] { x, y });
                    // Mark the neighboring cell as visited by updating the value in the grid
                    grid[x][y] = '0';
                }
            }
        }
    }
    /*
     * Time complexity: O(m*n) where m is the number of rows and n is the number of
     * columns in the grid.
     * 
     * Space complexity: O(min(m,n)) where m is the number of rows and n is the
     * number of columns in the grid.
     */
}

File: Graph/Traversals/NumberOfIslands.py
from typing import List
from collections import deque

class NumberOfIslands:
    def isValid(self, x, y, grid):
        return 0 <= x < len(grid) and 0 <= y < len(grid[0]) and grid[x][y] == '1'

    def numIslands(self, grid: List[List[str]]) -> int:
        directions = [(-1, 0), (1, 0), (0, 1), (0, -1)]
        islands = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.bfs(i, j, grid, directions)
                    islands += 1

        return islands

    def bfs(self, i, j, grid, directions):
        q = deque([(i, j)])

        while q:
            x, y = q.popleft()

            for dir in directions:
                nx, ny = x + dir[0], y + dir[1]

                if self.isValid(nx, ny, grid):
                    q.append((nx, ny))
                    grid[nx][ny] = '0'

# Example usage:
grid = [
    ['1', '1', '1', '1', '0'],
    ['1', '1', '0', '1', '0'],
    ['1', '1', '0', '0', '0'],
    ['0', '0', '0', '0', '0']
]

solution = NumberOfIslands()
result = solution.numIslands(grid)
print(result)
